In [ ]:
!pip install -q --no-deps --upgrade transformers datasets peft bitsandbytes lightning
!pip install -q --upgrade lightning pytorch-lightning lightning-utilities

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 411.1/411.1 kB 7.5 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.1/76.1 MB 21.3 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 819.0/819.0 kB 37.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.5 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.4 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 7.7 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 9.9 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 12.9 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 7.8 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 76.1 MB/s eta 0:00:00:00:0100:01
ERROR: pip's dependency r

In [ ]:
import warnings
warnings.filterwarnings("ignore")
warnings.filterwarnings("ignore", message=".*pad_token_id.*")

import os
import time
import torch
import numpy as np
import pandas as pd
import random
import matplotlib.pyplot as plt
import seaborn as sns

from torch.optim import AdamW
import torch.nn.functional as F
from torch.utils.data import DataLoader
from tqdm.auto import tqdm
from sklearn.metrics import accuracy_score, f1_score, classification_report
from sklearn.model_selection import train_test_split

from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline, BitsAndBytesConfig
from huggingface_hub import login
from datasets import Dataset

from peft import prepare_model_for_kbit_training, get_peft_model, LoraConfig, PeftModel

import lightning as L
from lightning.pytorch.callbacks import EarlyStopping, ModelCheckpoint

2025-05-19 13:45:33.230452: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1747662333.636662      35 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1747662333.758492      35 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


## Zero/One/Few-shot with our fine-tuned model

We took the small LLaMA model we had trained and repeated the zero/one/few-shot tests we had first tried with Gemini. Because the model learned from a strict prompt, we asked it new questions in that exact format: just the topic, the comment, and a request for one sentiment label.

In [ ]:
token = "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"
login(token=token)

In [ ]:
DATA_PATH = "/kaggle/input/inputtest/test_data.csv"
df = (pd.read_csv(DATA_PATH))

In [ ]:
torch.random.manual_seed(0)
model_name = "meta-llama/Llama-3.2-1B-Instruct"

SEED = 42
TRAIN_SIZE = 50
TEST_SIZE = 500
print(TEST_SIZE)

500


In [ ]:
# sampling

df = df.rename(columns={"input": "text", "output": "label"})
df_train = df.sample(TRAIN_SIZE, random_state=SEED).reset_index(drop=True)            # few-shot pool

df_rest = df.drop(df_train.index).reset_index(drop=True)
df_test = df_rest.sample(TEST_SIZE, random_state=SEED).reset_index(drop=True)
df_test["label"] = df_test["label"].str.lower().str.strip()

In [ ]:
class PromptBuilder:
    def __init__(self, df_pool: pd.DataFrame, k_few: int = 3):
        self.pool = df_pool.reset_index(drop=True)
        self.one  = self.pool.sample(1, random_state=SEED).reset_index(drop=True)
        self.few  = self.pool.sample(k_few, random_state=SEED+1).reset_index(drop=True)

    def _fmt(self, txt: str) -> str:
        return txt.replace("\n", " ")[:160]

    # zero-shot
    def zero(self, x: str, t: str) -> str:
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Below is an instruction that describes a task. "
            "ُTask: Write a response that classifies the sentiment of the given comment.\n\n"
            f"### Instruction:\nClassify the sentiment of the following comment regarding the topic: {t}.\n\n"
            f"### Input:\n{x}\n\n### Response:\n"
        )

    # one-shot
    def one_shot(self, x: str, t: str) -> str:
        ex = self.one.iloc[0]
        ex_txt, ex_lab, ex_top = self._fmt(ex.text), ex.label, self._fmt(ex.topic)
        # ex_txt, ex_lab, ex_top = "goodluck brother", "positive", "r/wallstreetbets"
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Below is an instruction that describes a task. "
            "Task: Write a response that classifies the sentiment of the given comment.\n\n"
            f"### Instruction:\nClassify the sentiment of the following comment regarding the topic: {t}.\n"
            "Your answer must be exactly one word: positive, negative, or neutral.\n\n"
            f"### Example:\nTopic: {ex_top}\nComment: {ex_txt}\nSentiment: {ex_lab}\n\n"
            f"### Input:\n{x}\n\n### Response:\n"
        )

    # few-shot
    def few_shot(self, x: str, t: str) -> str:
        shots = [
            f'x{i}: "{self._fmt(r.text)}"  (topic="{self._fmt(r.topic)}") → y{i} = {r.label}'
            for i, r in enumerate(self.few.itertuples(), start=1)
        ]
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Below is an instruction that describes a task. "
            "Write a response that classifies the sentiment of the given comment.\n\n"
            f"### Instruction:\nClassify the sentiment of the following comment regarding the topic: {t}.\n"
            "Your answer must be exactly one word: positive, negative, or neutral.\n\n"
            f"### Examples:\n{chr(10).join(shots)}\n\n"
            f"### Input:\n{x}\n\n### Response:\n"
        )

In [ ]:
def evaluate(builder_fn, test_df: pd.DataFrame, k: int | None = None, seed: int = 42) -> pd.DataFrame:
    label, pred = [], []
    if k is None or k > len(test_df):
        subset = test_df
    else:
        subset = test_df.sample(k, random_state=seed)
    for row in tqdm(subset.itertuples(), total=len(subset)):
        label.append(row.label.lower())
        prompt = builder_fn(row.text, row.topic)
        pred.append(extract_label(run_llama(prompt)))
    return pd.DataFrame({"label": label, "pred": pred})

In [ ]:
def extract_label(text: str) -> str:
    tok = text.lower().strip().split()[0].strip(",.")
    return tok if tok in {"positive", "negative", "neutral"} else "error"

In [ ]:
def run_llama(prompt):
    toks = tokenizer(prompt, return_tensors="pt").to(device)
    with torch.no_grad():
        out = model.generate(**toks, max_new_tokens=10, do_sample=False, pad_token_id=tokenizer.eos_token_id)
    generated = tokenizer.decode(out[0][toks["input_ids"].shape[-1]:], skip_special_tokens=True)
    return generated.strip()

In [ ]:
LORA_SAVE_DIR = "/kaggle/input/inputmodel"

bnb_cfg = BitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16, bnb_4bit_use_double_quant=True,)

tokenizer = AutoTokenizer.from_pretrained(LORA_SAVE_DIR, trust_remote_code=True)

base = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-3.2-1B-Instruct",
    device_map="auto",
    quantization_config=bnb_cfg,
    trust_remote_code=True,
)

model = PeftModel.from_pretrained(base, LORA_SAVE_DIR, is_trainable=False).eval()
device = model.device

config.json:   0%|          | 0.00/877 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.47G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/189 [00:00<?, ?B/s]

In [ ]:
builder = PromptBuilder(df_train, k_few=3)

In [ ]:
df_zero = evaluate(builder.zero, df_test, k=TEST_SIZE)
print(f"Zero-shot accuracy: {accuracy_score(df_zero.label, df_zero.pred):.3f}")
print(classification_report(df_zero.label, df_zero.pred))

  0%|          | 0/500 [00:00<?, ?it/s]

Zero-shot accuracy: 0.748
              precision    recall  f1-score   support

    negative       0.76      0.77      0.76       172
     neutral       0.73      0.83      0.78       245
    positive       0.82      0.45      0.58        83

    accuracy                           0.75       500
   macro avg       0.77      0.68      0.71       500
weighted avg       0.75      0.75      0.74       500



In [ ]:
# With explicit positive comment
df_one = evaluate(builder.one_shot, df_test, k=TEST_SIZE)
print(f"One-shot accuracy: {accuracy_score(df_one.label, df_one.pred):.3f}")
print(classification_report(df_one.label, df_one.pred))

  0%|          | 0/500 [00:00<?, ?it/s]

One-shot accuracy: 0.674
              precision    recall  f1-score   support

    negative       0.92      0.39      0.55       172
     neutral       0.64      0.88      0.74       245
    positive       0.61      0.65      0.63        83

    accuracy                           0.67       500
   macro avg       0.72      0.64      0.64       500
weighted avg       0.73      0.67      0.66       500



In [ ]:
# With random label comment
df_one = evaluate(builder.one_shot, df_test, k=TEST_SIZE)
print(f"One-shot accuracy: {accuracy_score(df_one.label, df_one.pred):.3f}")
print(classification_report(df_one.label, df_one.pred))

  0%|          | 0/500 [00:00<?, ?it/s]

One-shot accuracy: 0.674
              precision    recall  f1-score   support

    negative       0.66      0.73      0.70       172
     neutral       0.67      0.78      0.72       245
    positive       0.86      0.23      0.36        83

    accuracy                           0.67       500
   macro avg       0.73      0.58      0.59       500
weighted avg       0.70      0.67      0.65       500



In [ ]:
df_few = evaluate(builder.few_shot, df_test, k=TEST_SIZE)
print(f"Few-shot accuracy: {accuracy_score(df_few.label, df_few.pred):.3f}")
print(classification_report(df_few.label, df_few.pred))

  0%|          | 0/500 [00:00<?, ?it/s]

Few-shot accuracy: 0.694
              precision    recall  f1-score   support

    negative       0.80      0.66      0.72       172
     neutral       0.76      0.70      0.73       245
    positive       0.47      0.73      0.57        83

    accuracy                           0.69       500
   macro avg       0.67      0.70      0.67       500
weighted avg       0.72      0.69      0.70       500



## Test results

| Setting | Accuracy | Macro-F1 | Main observation |
|---------|:--------:|:--------:|------------------|
| **Zero-shot** | **0.75** | **0.71** | Works well without extra examples. |
| One-shot – **random** example | 0.67 | 0.59 | Model tilts toward the sample’s label. |
| One-shot – **positive** example | 0.67 | 0.64 | Big boost for *positive* class, others drop. |
| **Few-shot (3 mixed examples)** | 0.69 | 0.67 | Balances the bias, close to zero-shot scores. |


### What these numbers tell us؟

* **Zero-shot already strong:** 75 % accuracy shows the fine-tuned model understands the task with no extra help.  
* **One-shot is risky:** one example drags the model toward that label.  
  - **With a random example:** precision for positive jumps to 0.86, but recall drops to 0.23. The model overall macro-F1 falls.
  -  **With an explicitly positive example:** scores are more balanced, positive precision 0.61 and recall 0.65. Yet accuracy is still lower than zero-shot.

* **Few-shot fixes the deviation:** three diverse examples raise macro-F1 back to 0.70 and give the most even class balance, but still don't reach zero-shot accuracy.  
* **Small model = prompt-sensitive:** tiny wording or class imbalance in the examples quickly changes the outcome.


### Challenges we faced

- **Fixed prompt:** changing the prompt style even a little made the model perform worse.

- **One-shot bias:** with only one example, the model leans too much toward that example’s label.

- **Low positive recall:** even in few-shot mode, the model still misses many positive cases (recall ≈ 0.73), while negative and neutral do better.